app/lua53/host/luac.c

```diff
@@ -359,256 +269,286 @@ int main(int argc, char *argv[]) {
         lua_close(L);
         continue; /* and loop around once more simulating restart */
       }
-      char *err = strdup(lua_tostring(L, -1));
+      char *
       lua_close(L);
       fatal(err);
     }
     lua_close(L);
     break;
   }
+#endif
   return EXIT_SUCCESS;
 }
-
 /*
-** $Id: luac.c,v 1.76 2018/06/19 01:32:02 lhf Exp $
-** print bytecodes
-** See Copyright Notice in lua.h
+**
 */
 
+static int msghandler(lua_State *L) {
+  const char *msg = lua_tostring(L, 1);
+  if (msg == NULL) /* is error object not a string? */
+    msg = lua_pushfstring(L, "(error object is a %s value)", luaL_typename(L, 1));
+  luaL_traceback(L, L, msg, 1); /* append a standard traceback */
+  return 1; /* return the traceback */
+}
+
+static int execfile(lua_State *L, const char *name) {
+  int status = luaL_loadfile(L, name);
+  if (status == LUA_OK) {
+    int base = lua_gettop(L);
+    lua_pushcfunction(L, msghandler); /* push message handler */
+    lua_insert(L, base); /* put it under function and args */
+    status = lua_pcall(L, 0, 0, base);
+    lua_remove(L, base); /* remove message handler from the stack */
+  }
+  if (status != LUA_OK) {
+    const char *e = lua_tostring(L, -1);
+    if (!strncmp(e,"!LFS ",5))
+      lua_error(L);
+    fprintf(stderr, "%s: %s\n", PROGNAME, e);
+    lua_pop(L, 1); /* remove message */
+  }
+  return status;
+}
+
+/*
+ ** $Id: luac.c,v 1.76 2018/06/19 01:32:02 lhf Exp $
+ ** print bytecodes
+ ** See Copyright Notice in lua.h
+ */
+
 #include <ctype.h>
 #include <stdio.h>
 
-#define luac_c
-#define LUA_CORE
-
 #include "ldebug.h"
-#include "lobject.h"
 #include "lopcodes.h"
 
-#define VOID(p)		((const void*)(p))
-
-static void PrintString(const TString* ts)
-{
- const char* s=getstr(ts);
- size_t i,n=tsslen(ts);
- printf("%c",'"');
- for (i=0; i<n; i++)
- {
-  int c=(int)(unsigned char)s[i];
-  switch (c)
-  {
-   case '"':  printf("\\\""); break;
-   case '\\': printf("\\\\"); break;
-   case '\a': printf("\\a"); break;
-   case '\b': printf("\\b"); break;
-   case '\f': printf("\\f"); break;
-   case '\n': printf("\\n"); break;
-   case '\r': printf("\\r"); break;
-   case '\t': printf("\\t"); break;
-   case '\v': printf("\\v"); break;
-   default:	if (isprint(c))
-   			printf("%c",c);
-		else
-			printf("\\%03d",c);
+#define VOID(p)((const void *)(p))
+
+static void PrintString(const TString *ts) {
+  const char *s = getstr(ts);
+  size_t i, n = tsslen(ts);
+  printf("%c", '"');
+  for (i = 0; i < n; i++) {
+    int c = (int)(unsigned char) s[i];
+    switch (c) {
+    case '"':  printf("\\\""); break;
+    case '\\': printf("\\\\"); break;
+    case '\a': printf("\\a");  break;
+    case '\b': printf("\\b");  break;
+    case '\f': printf("\\f");  break;
+    case '\n': printf("\\n");  break;
+    case '\r': printf("\\r");  break;
+    case '\t': printf("\\t");  break;
+    case '\v': printf("\\v");  break;
+    default:
+      printf(isprint(c) ? "%c" : "\\%03d", c);
```

Comment:
While "obviously" equivalent, this still takes a moment to check, so I'd advocate either for reverting to match mainline or proposing this change to mainline for their inclusion.  Many of the following changes also have the same feel.

Reply:
I will post a general discussion in #3206.  Have read and review in that light.